design/01-proc-macro.md

```diff
@@ -0,0 +1,290 @@
+
+This document proposes an evolution of Cairo’s compiler plugins system. The primary motivation behind this proposal is to allow dynamic loading of compiler plugins per compilation unit. This document specifies syntax for the definition of procedural macros, a high-level view of their implementation in the compiler, and outlines how they interact with the compilation process.
+
+## Context
+
+### Rust procedural macros
+
+This post assumes reader knowledge on Rust's procedural macros and how they are implemented. A good walkthrough has been published some time ago on IntelliJ Rust blog:
+
+1. https://blog.jetbrains.com/rust/2022/03/18/procedural-macros-under-the-hood-part-i/
+2. https://blog.jetbrains.com/rust/2022/07/07/procedural-macros-under-the-hood-part-ii/
+
+### System programming primitives
+
+This post assumes reader knowledge on foreign function interface (FFI) mechanisms in modern operating systems.
+
+### Plugins API
+
+The Cairo compiler has a system of semantic plugins, that allow generating Cairo code with Rust code. The entry point of compiler plugins API is the [`MacroPlugin`] trait, which, via another mechanisms, boils down to allowing users to provide a callback, which:
+
+1. Takes a reference to compiler's syntax database ([`SyntaxGroup`], i.e. compiler state until right after parsing is done).
+2. Takes an [`ast::ModuleItem`](https://github.com/starkware-libs/cairo/blob/4821865770ac9e57442aef6f0ce82edc7020a4d6/crates/cairo-lang-syntax/src/node/ast.rs#L8718).
+3. Produces a struct that:
+    - Can remove the `ModuleItem`.
+    - Can contain Cairo code (as a virtual source file) to inject as a submodule.
+    - Emits compiler diagnostics.
+    - Can emit additional arbitrary data structure along generated file.
+
+Macro plugins are executed on each node in the Cairo AST.
+
+Custom plugins can be fed to the compiler via [`RootDatabaseBuilder::with_plugin_suite` API](https://github.com/starkware-libs/cairo/blob/8290898d602f87fb5bc3a18f28a42430cf6d3fdc/crates/cairo-lang-compiler/src/db.rs#L96). It takes a [`PluginSuite` object](https://github.com/starkware-libs/cairo/blob/8290898d602f87fb5bc3a18f28a42430cf6d3fdc/crates/cairo-lang-semantic/src/plugin.rs#L18-L25), which defines a set of objects implementing the [`MacroPlugin` trait](https://github.com/starkware-libs/cairo/blob/8290898d602f87fb5bc3a18f28a42430cf6d3fdc/crates/cairo-lang-defs/src/plugin.rs#L80).
+
+Scarb currently simply assumes a fixed, predefined plugin sets for `lib`, `starknet-contract` and `test` targets. This interface is not stable and is subject to changes in future compiler versions.
```

Comment:
That's not true anymore, because now we have builtin packages that provide plugins